plugins/producer_plugin/producer_plugin.cpp

```diff
@@ -2055,7 +2069,47 @@ void producer_plugin_impl::produce_block() {
       return sigs;
    } );
 
-   chain.commit_block();
+   // Notes:
+   // - possibly use `std::future` in the future:
+   //   - caller that gets the future can decide thread details.
+   //   - look into thread pooling.
+   // - `propose_constructed_block` can fail in it's current implementation.
+   // - `append_external_block` cannot fail in it's current implementation.
+   // - to get the watermark:
+   //   - call `get_watermark`.
+   // - to get the producer:
+   //   - call `hbs->get_scheduled_producer(block_time)`.
+   //   - or call `chain.head_block_state()` of which you might call `account_name controller::pending_block_producer() const`.
+   // - to get LIBid or LIBnumber:
+   //   - both can be found in controller.
+   //   - for this case do not get the chain last irreversible block number but get the LIB number from the head block state.
+   //   - get dpos irreversible block number
+   //   - block header state also has the producer.
+   //   - block header state also has the signed block as well.
+   // - it was talked about how another function could be added to controller for dealing with precursor cases prior to calling `commit_block`.
+   // - not sure why I wrote this function down but I'm keeping it around: `chain.fetch_block_by_number()`.
+   // - also keeping this note around `bsp = chain.finalize_block( [&]( const digest_type& d )` // get dpos irreversible block number
+
+   if ( blockvault_plug ) {
+      blockvault::watermark_t watermark{get_watermark(chain.head_block_state()->get_scheduled_producer(calculate_pending_block_time()).producer_name).value()};
+      uint32_t lib{chain.head_block_state()->dpos_irreversible_blocknum};
+      signed_block_ptr block{chain.head_block_state()->block};
+      std::function<void(bool)> handler{[this, &chain](bool b){
+         if ( b ) {
+           chain.commit_block();
+           block_state_ptr new_bs = chain.head_block_state();
+           ilog("Produced block ${id}... #${n} @ ${t} signed by ${p} [trxs: ${count}, lib: ${lib}, confirmed: ${confs}]",
+                 ("p",new_bs->header.producer)("id",new_bs->id.str().substr(8,16))
+                 ("n",new_bs->block_num)("t",new_bs->header.timestamp)
+                 ("count",new_bs->block->transactions.size())("lib",chain.last_irreversible_block_num())("confs", new_bs->header.confirmed));
+         } else {
+            _unapplied_transactions.add_aborted( chain.abort_block() );
+         }
+      }};
+      blockvault_plug->propose_constructed_block(watermark, lib, block, handler);
+   } else {
+      chain.commit_block();
+   }
 
    block_state_ptr new_bs = chain.head_block_state();
```

Comment:
If `proposed_constructed_block` determines that the block is no good then you should not log a "Produced block" here.

Reply:
`maybe_produce_block` should return `false` if unable to produce the block. It will return `true` even if the block is aborted now.

Reply:
For that matter this method should not return until the handler has been called because it will `schedule_production_loop` which will `start_block`. That can't happen until the current block is commit'ed or aborted.
This implies you need to pass a future into the handler and only continue once the future has been set. That will actually simplify this logic as you can just have the handler set the promise and then leave all the existing logic as is. The future->get() can either throw on unaccepted block or just return on accepted block.

Reply:
Implemented changes.